pkg/reconciler/cache/cachedresourceendpointsliceurls/cachedresourceendpointsliceurls_controller.go

```diff
@@ -0,0 +1,369 @@
+			_, err := clusterClient.CacheV1alpha1().CachedResourceEndpointSlices().Cluster(cluster).ApplyStatus(ctx, patch, metav1.ApplyOptions{
+				FieldManager: shardName,
+			})
+			return err
+		},
+	}
+
+	logger := logging.WithReconciler(klog.Background(), ControllerName)
+
+	_, _ = apiBindingInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueAPIBinding(objOrTombstone[*apisv1alpha2.APIBinding](obj), logger)
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueAPIBinding(objOrTombstone[*apisv1alpha2.APIBinding](newObj), logger)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueAPIBinding(objOrTombstone[*apisv1alpha2.APIBinding](obj), logger)
+		},
+	})
+
+	_, _ = localCachedResourceEndpointSliceClusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueCachedResourceEndpointSlice(objOrTombstone[*cachev1alpha1.CachedResourceEndpointSlice](obj), logger, "")
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueCachedResourceEndpointSlice(objOrTombstone[*cachev1alpha1.CachedResourceEndpointSlice](newObj), logger, "")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueCachedResourceEndpointSlice(objOrTombstone[*cachev1alpha1.CachedResourceEndpointSlice](obj), logger, "")
+		},
+	})
+
+	_, _ = globalCachedResourceEndpointSliceClusterInformer.Informer().AddEventHandler(events.WithoutSyncs(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueueCachedResourceEndpointSlice(objOrTombstone[*cachev1alpha1.CachedResourceEndpointSlice](obj), logger, " from cache")
+		},
+		UpdateFunc: func(_, newObj interface{}) {
+			c.enqueueCachedResourceEndpointSlice(objOrTombstone[*cachev1alpha1.CachedResourceEndpointSlice](newObj), logger, " from cache")
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.enqueueCachedResourceEndpointSlice(objOrTombstone[*cachev1alpha1.CachedResourceEndpointSlice](obj), logger, " from cache")
+		},
+	}))
+
+	return c, nil
+}
+
+func (c *controller) enqueueAPIBinding(obj *apisv1alpha2.APIBinding, logger logr.Logger) {
+	exportPath := logicalcluster.NewPath(obj.Spec.Reference.Export.Path)
+	if exportPath.Empty() {
+		exportPath = logicalcluster.From(obj).Path()
+	}
+	export, err := c.getAPIExport(exportPath, obj.Spec.Reference.Export.Name)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	logger = logging.WithObject(logger, obj)
+
+	for _, resource := range export.Spec.Resources {
+		if resource.Storage.Virtual == nil ||
+			ptr.Deref(resource.Storage.Virtual.Reference.APIGroup, "") != cachev1alpha1.SchemeGroupVersion.Group ||
+			resource.Storage.Virtual.Reference.Kind != "CachedResourceEndpointSlice" {
+			logger.V(4).Info("skipping APIBinding its referenced APIExport does not export CachedResourceEndpointSlice virtual resources")
```

Comment:
Does this mean we will not yet support other types of xEndpointSlices here?

Reply:
We do, but this is a CachedResourceEndpointSlice controller, and so it selects only exported resources with an object reference for this resource.